### ISSUE TITLE
Question about Sklearn Interface

### ISSUE THREAD

**Issue by nick-gorman** [quantumblacklabs/causalnex#75]
Hi, 

Thanks for this awesome package!

I wanted to get some more info on the Sklearn Interface.

How does the Sklearn Interface relate to structural models and Bayesian networks that can be used through other parts of the API?

Are the DAGRegressors just non probabilistic structural causal models where as Bayesian Networks are probabilistic? Is this the key difference?

Cheers,
Nick

**Comment by qbphilip** [comment]
Hey!

Thanks for your question.
How do you use the sklearn method? By itself or as first step to the probability fitting + "do" counterfactuals?

> How does the Sklearn Interface relate to structural models and Bayesian networks that can be used through other parts of the API?

The sklearn interface is a "supervised learning" wrapper for the structure learning step. It calls the same function but enforces that the target cannot be a parent. (i.e. we assume no reverse causality)

> Are the DAGRegressors just non probabilistic structural causal models where as Bayesian Networks are probabilistic? Is this the key difference?

You are right, the DAGClassifier/Regressor (or generally the structure learning) is non-probabilistic. Well, probabilistic but only in the sense that it is Likelihood based.

Currently (v0.8.1):
* Structure learning (structure or sklearn) is non-probabilistic but parametric (potentially nonlinear with pytorch)
* The Bayesian Network is fitting non-parametric probability distributions by binning continuous variables.


Since the summer we pushed for scaling the structure learning. A next step would be to update the probability fitting. I suppose that would need to be parametric, maybe via `pyro` instead of `pgmpy`? 

Another thought is to combine both steps and reusing a probabilistic structure model. Maybe via normalising flows but from what I understand right now is that they don't scale as well.

If you have suggestions, happy to hear! Contributions welcome as well.

Philip

**Comment by nick-gorman** [comment]
Hi Philip,

Thanks for your answer, I think that clarifies things.

I'm using the sklearn method by itself, by using the predict method for counterfactuals, i.e. what will Y be for a given value of x. Is that a valid use? I have continuous data, which I can't sensibly discretize, so I can't use the causalnex Bayesian networks.

This is not really my area so unfortunately I don't have much to contribute on those questions.

**Comment by pabloazurduy** [comment]
Hi guys 
I think that the documentation is not too clear about the `sklearn`tutorial and its difference with the "Bayesian network" implementation, also I realized that is possible to run `BayesianNetwork.fit_cpds(sm)` on top of a "non-discrete-`sm`" (what I mean by this is that the `dist_type_schema` contains `schema[col] = "cont"` elements). this was never explained in the documentation, but as I find out this scale very poorly (never finish the fitting). so I got stuck thinking... well what is the whole purpose of the `cont` distributions implementation ?, if it's not possible to fit a Bayesian network... or I might not understanding the connection between those two. 

(I'm not an expert on the subject, but seems like the tutorials [`sklearn` and `Distribution Schema`] need a little bit more information: 

- [ ] how to use a continuous distribution / non discrete `"sm"` after applying NOTEARS ? [what for ?](https://causalnex.readthedocs.io/en/latest/03_tutorial/dist_schema_tutorial.html)
- [ ] what is the relation between the `DAGRegressor` and `DAGClassificator` and the Bayesian implementation (the question on this topic )?

Thanks !!!

**Comment by rnrahman** [comment]
Hi! 

Thanks for the question. 

> I think that the documentation is not too clear about the sklearntutorial and its difference with the "Bayesian network" implementation, also I realized that is possible to run BayesianNetwork.fit_cpds(sm) on top of a "non-discrete-sm" (what I mean by this is that the dist_type_schema contains schema[col] = "cont" elements). this was never explained in the documentation, but as I find out this scale very poorly (never finish the fitting). so I got stuck thinking... well what is the whole purpose of the cont distributions implementation ?, if it's not possible to fit a Bayesian network... or I might not understanding the connection between those two.

We have explained this in our latest documentation. Current BN implementation will typically be badly fit and exhibit poor performance when structured model has continuous or any high cardinality features.
The purpose of having continuous variable as one of the distribution options is to give flexibility during structured learning process where having continuous variable helps. However, the reasons we didn't implement continuous in the probability fitting are a) distributions are rarely actually Gaussian b) discrete is much easier to work with c) mix of discrete and continuous is hard.

> how to use a continuous distribution / non discrete "sm" after applying NOTEARS ? [what for ?](https://causalnex.readthedocs.io/en/latest/03_tutorial/dist_schema_tutorial.html)
> what is the relation between the DAGRegressor and DAGClassificator and the Bayesian implementation (the question on this topic )?

* Not recommended to have continuous distribution for probability fitting since CausalNex's BN implementation can't handle high cardinality feature
* DAGRegressor and DAGClassifier are the sklearn wrappers for structured learning step. Bayesian network would be the second stage, after we define the structure of DAG. Bayesian network has CPD information in addition to DAG structure.


Cheers,
Ridlo
